File: jna-test/app/src/main/java/work/chiro/egui/Shader.java
package work.chiro.egui;

import com.google.common.io.Resources;

import java.io.IOException;
import java.nio.charset.Charset;

import static org.lwjgl.opengl.GL45.*;

public class Shader {
    public static final String VS_SRC_150;
    public static final String FS_SRC_150;
    public static final String VS_SRC;
    public static final String FS_SRC;

    static {
        try {
            VS_SRC_150 = Resources.toString(ClassLoader.getSystemResource("shaders/vs_150.shader"), Charset.defaultCharset());
            FS_SRC_150 = Resources.toString(ClassLoader.getSystemResource("shaders/fs_150.shader"), Charset.defaultCharset());
            VS_SRC = Resources.toString(ClassLoader.getSystemResource("shaders/vs.shader"), Charset.defaultCharset());
            FS_SRC = Resources.toString(ClassLoader.getSystemResource("shaders/fs.shader"), Charset.defaultCharset());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public static int compile(String src, int typ) {
        int shader = glCreateShader(typ);
        glShaderSource(shader, src);
        glCompileShader(shader);
        int[] status = new int[]{GL_FALSE};
        glGetShaderiv(shader, GL_COMPILE_STATUS, status);
        if (status[0] != GL_TRUE) {
            String log = glGetShaderInfoLog(shader);
            throw new RuntimeException(String.format("cannot compile shader: %s", log));
        }
        return shader;
    }

    public static int linkProgram(int vs, int fs) {
        int program = glCreateProgram();
        glAttachShader(program, vs);
        glAttachShader(program, fs);
        glLinkProgram(program);
        int[] status = new int[]{GL_FALSE};
        glGetProgramiv(program, GL_LINK_STATUS, status);
        if (status[0] != GL_TRUE) {
            String log = glGetProgramInfoLog(program);
            throw new RuntimeException(String.format("cannot link program: %s", log));
        }
        return program;
    }
}

File: cpp-opengl-test/Shader.cpp
//
// Created by chiro on 23-1-29.
//

#include "Shader.h"
#include <fmt/format.h>
#include <fstream>
#include <sstream>
#include "utils.h"
#include "gls.h"

Shader::Shader(const char *name) {
  std::string path_vert = fmt::format("../shaders/{}.vert", name);
  std::string path_frag = fmt::format("../shaders/{}.frag", name);
  if (!file_exists(path_vert) || !file_exists(path_frag)) {
    path_vert = fmt::format("shaders/{}.vert", name);
    path_frag = fmt::format("shaders/{}.frag", name);
    if (!file_exists(path_vert) || !file_exists(path_frag))
      throw std::runtime_error("file not found");
  }
  GLuint vert = Shader::compile(path_vert, GL_VERTEX_SHADER);
  GLuint frag = Shader::compile(path_frag, GL_FRAGMENT_SHADER);
  this->program = Shader::link(vert, frag);
}

GLuint Shader::compile(const std::string &path, int typ) {
  GLuint shader = glCreateShader(typ);
  std::ifstream file(path, std::ios::in);
  if (!file.is_open()) throw std::runtime_error("file not found");
  std::ostringstream stream;
  stream << file.rdbuf();
  std::string content = stream.str();
  glShaderSource(shader, 1, reinterpret_cast<const GLchar *const *>(content.c_str()), nullptr);
  glCompileShader(shader);
  int status;
  glGetShaderiv(shader, GL_COMPILE_STATUS, &status);
  if (status != GL_TRUE) {
    const GLsizei buf_size = 1024;
    GLsizei length = 0;
    GLchar buf[buf_size];
    glGetShaderInfoLog(shader, buf_size, &length, buf);
    buf[length] = '\0';
    throw std::runtime_error(buf);
  }
  return shader;
}

GLuint Shader::link(GLuint vs, GLuint fs) {
  GLuint program = glCreateProgram();
  glAttachShader(program, vs);
  glAttachShader(program, fs);
  glLinkProgram(program);
  int status;
  glGetProgramiv(program, GL_LINK_STATUS, &status);
  if (status != GL_TRUE) {
    const GLsizei buf_size = 1024;
    GLsizei length = 0;
    GLchar buf[buf_size];
    glGetProgramInfoLog(program, buf_size, &length, buf);
    buf[length] = '\0';
    throw std::runtime_error(buf);
  }
  return program;
}
